message/infrastructure/http/publisher.go

```diff
@@ -0,0 +1,104 @@
+package http
+
+import (
+	"bytes"
+	"net/http"
+
+	"github.com/pkg/errors"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+)
+
+var (
+	// ErrPublisherClosed happens when trying to publish to a topic while the publisher is closed or closing.
+	ErrPublisherClosed = errors.New("publisher is closed")
+)
+
+type MarshalMessageFunc func(topic string, msg *message.Message) (*http.Request, error)
+
+// DefaultMarshalMessageFunc encodes the UUID and Metadata in request headers
+// and sets the request URL according to the predefined server address and the topic.
+func DefaultMarshalMessageFunc(address string) MarshalMessageFunc {
+	return func(topic string, msg *message.Message) (*http.Request, error) {
+		req, err := http.NewRequest(http.MethodPost, address+"/"+topic, bytes.NewBuffer(msg.Payload))
+		if err != nil {
+			return nil, err
+		}
+
+		req.Header.Set(HeaderUUID, msg.UUID)
+
+		metadataJson, err := metadataToJson(msg)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not marshal metadata to JSON")
+		}
+		req.Header.Set(HeaderMetadata, string(metadataJson))
+		return req, nil
+	}
+}
+
+type Publisher struct {
+	client *http.Client
+	logger watermill.LoggerAdapter
+
+	marshalMessageFunc MarshalMessageFunc
+
+	closed bool
+}
+
+func NewPublisher(marshalMessageFunc MarshalMessageFunc, logger watermill.LoggerAdapter) (*Publisher, error) {
+	return NewPublisherWithClient(http.DefaultClient, marshalMessageFunc, logger)
+}
+
+func NewPublisherWithClient(client *http.Client, marshalMessageFunc MarshalMessageFunc, logger watermill.LoggerAdapter) (*Publisher, error) {
```

Comment:
If `DefaultMarshalMessageFunc` is not included in the default config, perhaps we should think of another name for it. Because it's not default then.

How about `HTTPPostRequestMarshalFunc` or something similar?

Reply:
Well, it will be still something "default" but not automatically set to the config. Eventually we can set it to the config by default, but in this case, we need to not forget to explicitly define this function in docs, etc. in other case users may not easily notice how to replace the default implementation

Reply:
afc279656462578257cb05721e69ef1835d5c3e3 adds configs for both Pub and Sub